tests/test_backend_general.py

```diff
@@ -0,0 +1,630 @@
+            "balance_transmission",
+            "flow_out_inc_eff",
+        ]
+        caplog.set_level(logging.DEBUG)
+
+        solved_model_func.backend.update_parameter("flow_out_eff", updated_param)
+
+        assert (
+            f"The optimisation problem components {refs_to_update} will be re-built."
+            in caplog.text
+        )
+
+        expected = solved_model_func.backend.get_parameter(
+            "flow_out_eff", as_backend_objs=False
+        )
+        assert "timesteps" in expected.dims
+
+    def test_update_parameter_replace_undefined(self, caplog, solved_model_func):
+        """source_eff isn't defined in the inputs, so is a dimensionless value in the pyomo object, assigned its default value."""
+        updated_param = solved_model_func.inputs.flow_out_eff
+
+        refs_to_update = ["balance_supply_no_storage"]
+        caplog.set_level(logging.DEBUG)
+
+        solved_model_func.backend.update_parameter("source_eff", updated_param)
+
+        assert (
+            f"The optimisation problem components {refs_to_update} will be re-built."
+            in caplog.text
+        )
+
+        expected = solved_model_func.backend.get_parameter(
+            "source_eff", as_backend_objs=False
+        )
+        assert expected.equals(updated_param)
+
+    @pytest.mark.parametrize("model_suffix", ["_longnames", "_updated_cost_flow_cap"])
+    @pytest.mark.parametrize(
+        ("expr", "kwargs"),
+        [
+            ("cost_investment_flow_cap", {"carriers": "electricity"}),
+            ("cost_investment", {}),
+            ("cost", {}),
+        ],
+    )
+    @pytest.mark.usefixtures(
+        "built_model_func_longnames", "built_model_func_updated_cost_flow_cap"
+    )
+    def test_update_parameter_expr_refs_rebuilt(
+        self, request: pytest.FixtureRequest, model_suffix: str, expr: str, kwargs: dict
+    ):
+        """Check that parameter re-definition propagates across all cross-referenced global expressions."""
+        model: calliope.Model = request.getfixturevalue(
+            "built_model_func" + model_suffix
+        )
+        expression_string = (
+            model.backend.get_global_expression(expr, as_backend_objs=False)
+            .sel(techs="test_demand_elec", **kwargs)
+            .astype(str)
+        )
+        if model_suffix.endswith("updated_cost_flow_cap"):
+            assert expression_string.str.contains("test_demand_elec").all()
+        else:
+            assert not (expression_string.str.contains("test_demand_elec").any())
+
+    @pytest.mark.usefixtures(
+        "built_model_func_longnames", "built_model_func_updated_cost_flow_cap"
+    )
+    @pytest.mark.parametrize("model_suffix", ["_longnames", "_updated_cost_flow_cap"])
+    def test_update_parameter_refs_in_obj_func(
+        self, request: pytest.FixtureRequest, model_suffix: str
+    ):
+        """Check that parameter re-definition propagates from global expressions to objective function."""
```

Comment:
Nice-to-have: I think only checking the updated version is sufficient here. The other seems a bit like "testing your own test". LMKWYT.

Reply:
hmm, I guess I want to catch false negatives in future here (if we give `test_demand_elec` a value for some reason, the `updated` check will pass but not necessarily because the functionality we're testing is passing.